rllib/utils/replay_buffers/reservoir_buffer.py

```diff
@@ -0,0 +1,109 @@
+from typing import Any, Dict
+import random
+
+# Import ray before psutil will make sure we use psutil's bundled version
+import ray  # noqa F401
+import psutil  # noqa E402
+
+from ray.rllib.utils.annotations import ExperimentalAPI, override
+from ray.rllib.utils.replay_buffers.replay_buffer import ReplayBuffer
+from ray.rllib.utils.typing import SampleBatchType
+from ray.rllib.execution.buffers.replay_buffer import warn_replay_capacity
+
+
+@ExperimentalAPI
+class ReservoirBuffer(ReplayBuffer):
+    """This buffer implements reservoir sampling.
+
+    The algorithm has been described by Jeffrey S. Vitter in "Random sampling
+    with a reservoir". See https://www.cs.umd.edu/~samir/498/vitter.pdf for
+    the full paper.
+    """
+
+    def __init__(self, capacity: int = 10000, storage_unit: str = "timesteps"):
+        """Initializes a ReservoirBuffer instance.
+
+        Args:
+            capacity: Max number of timesteps to store in the FIFO
+                buffer. After reaching this number, older samples will be
+                dropped to make space for new ones.
+            storage_unit: Either 'sequences' or 'timesteps'. Specifies how
+                experiences are stored.
+        """
+        ReplayBuffer.__init__(self, capacity, storage_unit)
+        self._num_add_calls = 0
+        self._num_evicted = 0
+
+    @ExperimentalAPI
+    @override(ReplayBuffer)
+    def add(self, batch: SampleBatchType, **kwargs) -> None:
+        """Adds a batch of experiences.
+
+        Args:
+            batch: SampleBatch to add to this buffer's storage.
+        """
+        # Update add counts.
+        self._num_add_calls += 1
+        # Update our timesteps counts.
+        self._num_timesteps_added += batch.count
+        self._num_timesteps_added_wrap += batch.count
+
+        if self._num_timesteps_added < self.capacity:
+            ReplayBuffer.add(self, batch)
+        else:
+            # Eviction of older samples has already started (buffer is "full")
+            self._eviction_started = True
+            idx = random.randint(0, self._num_add_calls - 1)
+            if idx < self.capacity:
+                self._num_evicted += 1
+                self._evicted_hit_stats.push(self._hit_count[idx])
```

Comment:
you don't want to simply save the distribution _hit_count?
the list could grow pretty large over time, this is almost like a memory leak ... :)

Reply:
Hi! I have not looked into the implementation of WindowStat (=type(self._evicted_hit_stats)), and I will eventually, but WindowStat should drop old values as new ones are pushed, right? I was guessing only the last n values are of concern to such a statistic. The line occurs in the old ReplayBuffer class and I am yet to double check it.